### ISSUE TITLE
Case clauses should be able to match list indexes

### ISSUE THREAD

**Issue by CodesMaha** [python/cpython#128313]
# Feature or enhancement

> [!NOTE]
> (by @picnixz): Feature is under DPO consideration. Issue will be re-opened once sufficient support has been gathered (and likely a PEP has been accepted).

### Proposal:

While list indexes can be written to match to other variables or constants, list indexes unfortunately currently cannot be cases to be matched with. Additionally, this dynamic sort of feature also already exists in JavaScript; although, it actually uses simpler value matching instead of languages like Python or C# that make use of structural pattern matching.

If addressed, Python developers will be able to simply edit a list to change case clauses without having to scroll through each and every one to find the one they'd like to change.

The use could potentially resemble the following example code…

```python
lst = ['a', 'b', 'c'] # of options in this example

user_input = input(f'Input {lst[0]}, {lst[1]}, or {lst[2]}. \n: ').strip().lower()

match user_input:

    case lst[0]:
        print('\nFirst option chosen. ')
        ...

    case lst[1]:
        print('\nSecond option chosen. ')
        ...

    case lst[2]:
        print('\nThird option chosen. ')
        ...

    case _:
        print('\nInvalid input. Kindly try again.')
        ...

# shows a SyntaxError
```

However, as mentioned here later, someone suggested an edit in the above example code and provided the following code block — with the difference being the addition of `==`, which can be `is` instead if appropriate…

```python
lst = ['a', 'b', 'c'] # of options in this example

user_input = input(f'Input {lst[0]}, {lst[1]}, or {lst[2]}. \n: ').strip().lower()

match user_input:
    case == lst[0]:
        print('\nFirst option chosen. ')
        ...
    case == lst[1]:
        print('\nSecond option chosen. ')
        ...
    case == lst[2]:
        print('\nThird option chosen. ')
        ...
    case _:
        print('\nInvalid input. Kindly try again.')
        ...
```

They also highlighted how this addition had "an obvious (and accurate) mapping" to a version of the example code using `if`-`else` statements, for which they provided the following edition continuing after declarations…

```python
if user_input == lst[0]:
    print('\nFirst option chosen. ')
    ...
elif user_input == lst[1]:
    print('\nSecond option chosen. ')
    ...
elif user_input == lst[2]:
    print('\nThird option chosen. ')
    ...
else:
    print('\nInvalid input. Kindly try again.')
    ...
```

### Has this already been discussed elsewhere?

No response given

### Links to previous discussion of this feature:

LinkedIn Post: [https://www.linkedin.com/posts/profile-of-maha-ijaz_serious-careerchoice-humour-activity-7278731541126524929-6jfI?utm_source=share](https://www.linkedin.com/posts/profile-of-maha-ijaz_serious-careerchoice-humour-activity-7278731541126524929-6jfI?utm_source=share)

The image attached to the post describes how JavaScript already has this feature in its switch-case statements while the latest stable version of Python (3.13.x) does not.

---

Python discussion: [https://discuss.python.org/t/case-clauses-should-be-able-to-match-list-indexes/75540](https://discuss.python.org/t/case-clauses-should-be-able-to-match-list-indexes/75540)

Opened after issue — sorry.

**Summary of state as of December 30, 2024**:
+ Two links presented, one to this same GitHub (GH) issue and another to the Stack Overflow (SO) question which is mentioned here afterwards as well.
+ A core CPython developer expressed approval and also suggested a subfeature from [PEP 642](https://peps.python.org/pep-0642/), which that same person authored. The subfeature in question allows for `==` and `is` to be used as prefix operations to replace implicit value patterns in this case.

---

Old but related SO question: 
[https://stackoverflow.com/questions/66159432/how-to-use-values-stored-in-variables-as-case-patterns](https://stackoverflow.com/questions/66159432/how-to-use-values-stored-in-variables-as-case-patterns)

In this, one account even said, 
> that's been proposed for Python (and rejected) twice before …

It's interesting to see how many times this exact topic has been brought up or revisited.

**Comment by ZeroIntensity** [comment]
Please bring this up on [DPO](https://discuss.python.org/c/ideas/6) first, but in general, this isn't what `match` is supposed to be for--you should just use an `if` here (or better yet, a dictionary mapping of choices to functions). `match` is for *structural pattern matching*, and this isn't structural nor a pattern. You might want to take a look at PEP-636 for the use cases.

**Comment by picnixz** [comment]
In this case, you can use a `StrEnum`. The issue with `lst[0]` is that the latter is not a constant.

**Comment by CodesMaha** [comment]
> Please bring this up on [DPO](https://discuss.python.org/c/ideas/6) first,

Of course, done :)
Sorry for not doing so earlier.

No one has replied yet, though, so I'm not sure if the general consensus is with or against this on there.

> you should just use an `if` here 

Yes, that is what I did in my personal Python code a few days ago when I discovered this before researching it. 
Thank you for your suggestion nonetheless.

> `match` is for _structural pattern matching_, and this isn't structural nor a pattern. You might want to take a look at [PEP-636](https://peps.python.org/636) for the use cases.

Sure, I can definitely do that again if you insist :)

**Comment by ZeroIntensity** [comment]
> No one has replied yet, though, so I'm not sure if the general consensus is with or against this on there.

A lot of people (especially core devs) are off for the holidays at the moment, so I wouldn't expect speedy replies.

**Comment by CodesMaha** [comment]
> The issue with `lst[0]` is that the latter is not a constant (or more precisely, it cannot be understood correctly in a structural pattern matching context)

Yes, and I saw that the reason it worked for JavaScript is because it used simpler value matching, whereas languages like Python or C# use the aforementioned structural pattern matching.

I'm not sure if this issue should be closed then, though.

> In this case, you can use a `StrEnum`

I agree; that is indeed a good alternative for the meanwhile for anyone encountering this.

Thank you for your comment :)

**Comment by picnixz** [comment]
In order to clean the `pending` backlog, I'm closing this issue as `not planned` for now since the DPO thread did not have a consensus (last message was 2 weeks ago). Nonetheless, if the idea is revived, just ping me on DPO and we'll re-open the issue (I'm `@picnixz` out there as well).
